README.md
# Ronin Gen

* http://ronin.rubyforge.org/gen/
* http://github.com/postmodern/ronin-gen
* http://github.com/postmodern/ronin-gen/issues
* http://groups.google.com/group/ronin-ruby
* irc.freenode.net #ronin

## DESCRIPTION:

Ronin Gen is a Ruby library for Ronin that provides various generators.

Ronin is a Ruby platform for exploit development and security research.
Ronin allows for the rapid development and distribution of code, exploits
or payloads over many common Source-Code-Management (SCM) systems.

### Ruby

Ronin's Ruby environment allows security researchers to leverage Ruby with
ease. The Ruby environment contains a multitude of convenience methods
for working with data in Ruby, a Ruby Object Database, a customized Ruby
Console and an extendable command-line interface.

### Extend

Ronin's more specialized features are provided by additional Ronin
libraries, which users can choose to install. These libraries can allow
one to write and run Exploits and Payloads, scan for PHP vulnerabilities,
perform Google Dorks  or run 3rd party scanners.

### Publish

Ronin allows users to publish and share code, exploits, payloads or other
data via Overlays. Overlays are directories of code and data that can be
hosted on any SVN, Hg, Git or Rsync server. Ronin makes it easy to create,
install or update Overlays.

## FEATURES:

* Provides {Ronin::Generators::Generator}, a Thor based generator class that
  can be used to create new generators.
* Provides {Ronin::Generators::DirGenerator}, a Thor based generator that
  can be used to create new directory generators.
* Generators for Overlays and Extensions.

## SYNOPSIS:

Generate a skeleton Overlay:

    $ ronin-gen overlay PATH [options]

Generate a skeleton Extension:

    $ ronin-gen extension PATH

## REQUIREMENTS:

* [ronin](http://ronin.rubyforge.org/) >= 0.3.0

## INSTALL:

    $ sudo gem install ronin-gen

## LICENSE:

Ronin Gen - A Ruby library for Ronin that provides various generators.

Copyright (c) 2009-2010 Hal Brodigan (postmodern.mod3 at gmail.com)

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA


History.md
### 0.2.0 / 2009-09-24

* Require ronin >= 0.3.0.
* Require rspec >= 1.2.8.
* Require yard >= 0.2.3.5.
* Updated the project summary and 3-point description of Ronin Gen.
* Moved to YARD based documentation.
* Refactored Ronin::Generators::Generator to inherit from Thor.
* Refactored Ronin::Generators::DirGenerator.
* Refactored Ronin::Generators::Platform::Extension.
* Refactored Ronin::Generators::Platform::Overlay.
* Rewrote the Ronin::UI::CommandLine generator commands to simply inherit
  from the new Thor based generator classes within Ronin::Generators.

### 0.1.1 / 2009-07-02

* Use Hoe >= 2.0.0.
* Require Ronin >= 0.2.4.
* Added Generator#touch.
* Added Ronin::UI::CommandLine::Commands::Gen for listing and
  invoking generators.
* Added more defaults to Ronin::Generators::Platform::Overlay:
  * Default maintainer.
  * Default license.
  * Default description.
* Added the --version option to the ronin-gen command.
* Removed the --list option from the ronin-gen command.
* Make sure Ronin::Generators::Platform::Overlay creates an
  empty lib/init.rb file.
* Fixed a bug in the ronin-gen command, where generator names
  were not being sanitized.

### 0.1.0 / 2009-03-27

* Initial release.
  * Provides the Ronin::Generators::Platform::Overlay generator.
  * Provides the Ronin::Generators::Platform::Extension generator.
  * Provides sub-commands for invoking the Overlay and Extension generator.
  * Provides the ronin-gen sub-command to invoke other generator
    sub-commands.



lib/ronin/gen/library.rb
#
# Ronin Gen - A Ruby library for Ronin that provides various generators.
#
# Copyright (c) 2009-2010 Hal Brodigan (postmodern.mod3 at gmail.com)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#

require 'ronin/gen/dir_generator'
require 'ronin/gen/config'
require 'ronin/version'

module Ronin
  module Gen
    class Library < DirGenerator

      DEFAULT_VERSION = '0.1.0'

      DEFAULT_AUTHOR = 'Author'

      DEFAULT_EMAIL = 'name@host.com'

      RSPEC_VERSION = '1.3.0'

      YARD_VERSION = '0.5.3'

      class_option :name, :type => :string
      class_option :author, :type => :string, :default => DEFAULT_AUTHOR
      class_option :email, :type => :string, :default => DEFAULT_EMAIL
      class_option :version, :type => :string, :default => DEFAULT_VERSION

      def setup
        @name = (options[:name] || File.basename(self.destination_root))
        @dir_name = @name.gsub(/^ronin[-_]/,'')
        @module_name = @dir_name.capitalize

        @title = @name.split(/[\s_-]+/).map { |word|
          word.capitalize
        }.join(' ')

        @author = options[:author]
        @email = options[:email]
        @safe_email = @email.gsub(/\s*@\s*/,' at ')
      end

      def generate
        copy_file File.join('ronin','gen','library','COPYING.txt'),
                  'COPYING.txt'

        mkdir 'static'
      end

      def bin
        mkdir 'bin'
        template File.join('ronin','gen','library','bin','ronin-example.erb'),
                 File.join('bin',"ronin-#{@dir_name}")
      end

      def lib
        mkdir File.join('lib','ronin',@dir_name)

        template File.join('ronin','gen','library','lib','ronin','example.rb.erb'),
                 File.join('lib','ronin',"#{@dir_name}.rb")
        template File.join('ronin','gen','library','lib','ronin','example','version.rb.erb'),
                 File.join('lib','ronin',@dir_name,'version.rb')
      end

      def history
        template File.join('ronin','gen','library','History.md.erb'),
                 'History.md'
      end

      def manifest
        template File.join('ronin','gen','library','Manifest.txt.erb'),
                 'Manifest.txt'
      end

      def readme
        template File.join('ronin','gen','library','README.md.erb'),
                 'README.md'
      end

      def rakefile
        template File.join('ronin','gen','library','Rakefile.erb'),
                 'Rakefile'
      end

      def test_suite
        mkdir 'spec'
        template File.join('ronin','gen','library','spec','spec_helper.rb.erb'),
                 File.join('spec','spec_helper.rb')

        mkdir File.join('spec',@dir_name)
        template File.join('ronin','gen','library','spec','example','example_spec.rb.erb'),
                 File.join('spec',@dir_name,"#{@dir_name}_spec.rb")
      end

    end
  end
end
